intermediate_source/torch_export_tutorial.py

```diff
@@ -0,0 +1,462 @@
+# are necessary. We can do this by relaxing all constraints and letting ``torch.export``
+# error out.
+
+inp4 = torch.randn(8, 16)
+inp5 = torch.randn(16, 32)
+
+def constraints_example3(x, y):
+    if x.shape[0] <= 16:
+        return x @ y[:, :16]
+    return y
+
+constraints3 = (
+    [dynamic_dim(inp4, i) for i in range(inp4.dim())] +
+    [dynamic_dim(inp5, i) for i in range(inp4.dim())]
+)
+
+try:
+    export(constraints_example3, (inp4, inp5), constraints=constraints3)
+except Exception:
+    tb.print_exc()
+
+def specify_constraints(x, y):
+    return [
+        # x:
+        dynamic_dim(x, 0),
+        dynamic_dim(x, 1),
+        dynamic_dim(x, 0) <= 16,
+
+        # y:
+        16 < dynamic_dim(y, 1),
+        dynamic_dim(y, 0) == dynamic_dim(x, 1),
+    ]
+
+constraints3_fixed = specify_constraints(inp4, inp5)
+exported_constraints_example3 = export(constraints_example3, (inp4, inp5), constraints=constraints3_fixed)
+print(exported_constraints_example3(torch.randn(4, 32), torch.randn(32, 64)))
+
+######################################################################
+# Note that in the example above, because we constrained the value of ``x.shape[0]`` in
+# ``constraints_example3``, the exported program is sound even though there is a
+# raw ``if`` statement.
+#
+# If you want to see why ``torch.export`` generated these constraints, you can
+# re-run the script with the environment variable ``TORCH_LOGS=dynamic,dynamo``,
+# or use ``torch._logging.set_logs``.
+
+import logging
+torch._logging.set_logs(dynamic=logging.INFO, dynamo=logging.INFO)
+exported_constraints_example3 = export(constraints_example3, (inp4, inp5), constraints=constraints3_fixed)
+torch._logging.set_logs(dynamic=logging.WARNING, dynamo=logging.WARNING)
+
+######################################################################
+# We can view an ``ExportedProgram``'s constraints using the ``range_constraints`` and
+# ``equality_constraints`` attributes. The logging above reveals what the symbols ``s0, s1, ...``
+# represent.
+
+print(exported_constraints_example3.range_constraints)
+print(exported_constraints_example3.equality_constraints)
+
+######################################################################
+# We can also constrain on individual values in the source code itself using
+# ``constrain_as_value`` and ``constrain_as_size``. ``constrain_as_value`` specifies
+# that a given integer value is expected to fall within the provided minimum/maximum bounds (inclusive).
+# If a bound is not provided, then it is assumed to be unbounded. ``constrain_as_size``
+# is similar to ``constrain_as_value``, except that it should be used on integer values that
+# will be used to specify tensor shapes -- in particular, the value must not be 0 or 1 because
+
+from torch.export import constrain_as_size, constrain_as_value
+
+def constraints_example4(x, y):
+    b = y.item()
+    constrain_as_value(b, 3, 5)
```

Comment:
What happens if the constraint is not true for some input? Is it an exception at export time, or would you get it later?

Reply:
If the inputs to export violate the constraints, the error happens at export time (example will be included). If future inputs violate constraints, the error happens when you try to run the exported graph on the future inputs.